packages/next/src/build/transpile-config.ts

```diff
@@ -0,0 +1,169 @@
+import { dirname, join, resolve } from 'path'
+import { pathToFileURL } from 'url'
+import { runCompiler } from './compiler'
+import { ProfilingPlugin } from './webpack/plugins/profiling-plugin'
+import { JsConfigPathsPlugin } from './webpack/plugins/jsconfig-paths-plugin'
+import { trace } from '../trace'
+import { hasNecessaryDependencies } from '../lib/has-necessary-dependencies'
+import { getTypeScriptConfiguration } from '../lib/typescript/getTypeScriptConfiguration'
+import type { ParsedCommandLine } from 'typescript'
+import type { webpack } from 'next/dist/compiled/webpack/webpack'
+import type { SWCLoaderOptions } from './webpack/loaders/next-swc-loader'
+import type { NextConfig } from '../server/config-shared'
+
+export async function transpileConfig({
+  configPath,
+  configFileName,
+  cwd,
+}: {
+  configPath: string
+  configFileName: string
+  cwd: string
+}): Promise<NextConfig> {
+  const isCJS = configFileName.endsWith('.cts')
+  const filename = `next.compiled.config.${isCJS ? 'cjs' : 'mjs'}`
+
+  let tsConfig: ParsedCommandLine
+  try {
+    const deps = await hasNecessaryDependencies(cwd, [
+      {
+        pkg: 'typescript',
+        file: 'typescript/lib/typescript.js',
+        exportsRestrict: true,
+      },
+    ])
+    const typeScriptPath = deps.resolved.get('typescript')
+
+    const ts = (await Promise.resolve(
+      require(typeScriptPath!)
+    )) as typeof import('typescript')
+
+    tsConfig = await getTypeScriptConfiguration(
+      ts,
+      join(cwd, 'tsconfig.json'),
+      true
+    )
+  } catch {
+    tsConfig = { options: {} } as ParsedCommandLine
+  }
+
+  async function bundleConfig(nextConfig: NextConfig, inputFileSystem?: any) {
```

Comment:
Why do we need to bundle here, could we instead just transpile on-demand instead so we do the least amount of work necessary?

Reply:
To resolve imports from other ts files, as far as I tried, swc transform didn't support that.

Reply:
How about we introduce an option like "preCompiled" on the plugins of tsconfig.json, and let users list the files that are imported in the next.config.ts file? This will ensure much less perf drop and complexity.

```json
{
  "compilerOptions": {
    "plugins": [
      {
        "name": "next",
        "preCompiled": ["./foo", "./bar"]
      }
    ]
  }
}
```

Reply:
We could transpile to CJS and then use a require hook to handle transpiling children modules from `next.config.ts`

Reply:
@ijjk Having an issue with SWC resolving `paths` and `baseUrl`, but seems like most of the cases are handled at the moment. Could you confirm that this kind of require-hook is what you were expecting?